File: Python/DynamicProgramming/LIS.py
def lengthOfLIS(nums):
    n = len(nums)

    def helper(curr_index, prev_index):
        # base case
        if curr_index >= n:
            return 0

        # exclude current element
        exclude = helper(curr_index + 1, prev_index)

        # include current element
        include = 0
        if prev_index == -1 or nums[curr_index] > nums[prev_index]:
            include = 1 + helper(curr_index + 1, curr_index)

        return max(exclude, include)

    return helper(0, -1)

File: Golang/DynamicProgramming/LIS.go
package main

func lengthOfLIS(nums []int) int {
	n := len(nums)

	var helper func(int, int) int
	helper = func(currIndex int, prevIndex int) int {
		// base case
		if currIndex >= n {
			return 0
		}

		// exclude
		exclude := helper(currIndex+1, prevIndex)

		// include
		include := 0
		if prevIndex == -1 || nums[currIndex] > nums[prevIndex] {
			include = 1 + helper(currIndex+1, currIndex)
		}

		if include > exclude {
			return include
		}
		return exclude
	}

	return helper(0, -1)
}
